### DISCUSSION TITLE
Help Error: [mobx-state-tree] assertion failed: cannot finalize the creation of a node that is already dead

### DISCUSSION THREAD

**Discussion by eldawwy** [mobxjs/mobx-state-tree#2225] [Q&A] [answered]
I am encountering an issue with mobx-state-tree when trying to manage a list of running orders in my application. The error occurs under the following circumstances:

`[Error: [mobx-state-tree] assertion failed: cannot finalize the creation of a node that is already dead]`
The error appears when attempting to delete an order from the running array using the destroy(order) function inside the deleteRunningOrder action.
The deletion logic is designed to remove an order when certain conditions are met (e.g., the order has expired).

<img width="665" alt="Screenshot 2024-11-24 at 11 38 44 PM" src="https://github.com/user-attachments/assets/93c8c1eb-6e84-44eb-9b55-49c390af3d79">
<img width="989" alt="Screenshot 2024-11-24 at 11 39 25 PM" src="https://github.com/user-attachments/assets/97449e15-a789-44ab-a7f2-33293e528ff5">

**Comment by coolsoftwaretyler** [comment]
Hey @xforcarlos - sorry to hear about the trouble. My guess is you've got some kind of asynchronous operations happening somewhere that don't settle before you modify them. Since you're calling `setProp` on the `running` array in `Orders` - I'm guessing that you begin to assign some nodes, then maybe assign a new set of them from an API response, and the first set of nodes hit the `afterCreate` hook after they've been swapped out.

That's just a guess though. I could help you a lot more if you're able to create a minimal reproduction? That would be clearer than screenshots of the code, and you might even be able to debug your own issue by trying to take your app logic and distilling it to the smallest possible example where you can reproduce the issue.

Looking forward to helping out!

**Reply by eldawwy** [reply]
Hello @coolsoftwaretyler 
thanks for quick answer 
i post a screenshot because code snippet not format the code 
 and i can't reproduction because the model is complicated but the let me explain this

i fetch list of orders and there  is socket too for realtime & on orders model add new order

the idea is delete the order from list if the autoAssignedAt + waitTime end  
and i have another login delete order 
like when status update or driver update by button and all is work fine except this

**Reply by coolsoftwaretyler** [reply]
Hi @xforcarlos - I totally understand. I still recommend putting together a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) as an exercise when asking for help about code. You may find that the process of distilling your example into something that reproduces the error helps you figure out your own problem.

@thegedge has another good possibility for you in a separate comment. Let us know if that helps.

**Reply by eldawwy** [reply]
@coolsoftwaretyler 
thanks for help i couldn't provide minimal reproducible example
because it's too complicated logic

**Comment by thegedge** [comment, marked as answer]
I think it's the `self.deleteOrder` method you have in `afterCreate` (the one that's not in `setTimeout`).

The `afterCreate` is fired before a node is "finalized" (more of an internal detail, but hopefully that brings clarity to the error message). The problem is that it's getting `destroy`-ed before that happens, and hence the error message.

I think you could resolve this by doing the `deleteOrder` call in `setTimeout(..., 0)`, because it will run (roughly) on the next tick of the event loop, after the object has had a chance to finalize.

**Reply by eldawwy** [reply]
@thegedge thanks Bro for help
i was have another afterCreate method that's handle another logic

so i merge both and solved 

` afterCreate() {
      const setOrderChange = ({ order }: { order: OrderSnapshotOut }) => {
        const { authenticationStore } = getRootStore(self)
        const orderSnapShot = convertForTestOrder(order)
        if (deleteOrderStatus.includes(OrderStatus[orderSnapShot.currentStatus])) {
          this.handleCanceled(orderSnapShot)
          return
        }
        if (acceptedOrderStatus.includes(OrderStatus[orderSnapShot.currentStatus])) {
          this.handleAccepted(orderSnapShot)
          return
        }
        if (order?.driver && order?.driver?._id !== authenticationStore.user?._id) {
          this.handleCanceled(orderSnapShot)
          return
        }
        applySnapshot(self, orderSnapShot)
      }
      if (self.isPastOrder || !self._id) return
      self.realTime.onConnect(() => {
        __DEV__ && console.tron.log("order_Id-----onConnect", isAlive(self) && self._id)
      })
      self.realTime.onEmit("join-room", { room: self._id })
      if (!self.orderChangeListener) {
        self.orderChangeListener = setOrderChange
        self.realTime.onChange(self._id, self.orderChangeListener)
      }
      if (self.autoAssignedAt !== null) {
        const { watiSec } = getRootStore(self).appSettingStore
        const autoAssignedTime = new Date(self.autoAssignedAt).getTime()
        const expiryTime = autoAssignedTime + watiSec * 1000
        const delay = expiryTime - Date.now()
        if (delay > 0) {
          self.timeoutId = setTimeout(() => {
            self.deleteOrder()
          }, delay)
          console.log("delay", delay)
        } else {
          console.log("TEST")
          self.deleteOrder()
        }
      }
    },
    beforeDestroy() {
      if (self.isPastOrder) return
      self.realTime.remove(self._id, self.orderChangeListener)
      clearInterval(self.timeoutId)
    },`
